### PR TITLE
*: print the statement causing write conflicts

### LINKED ISSUE #14593: Print the SQL causing write conflict in logs
## Feature Request

Now when write conflicts occur in optimistic transactions, only the last SQL is printed out in logs. However, the last SQL statement is not always the one which causes conflicts. It's better to print the exact statement which is related to the conflicts.

Example:

```
MySQL [test]> select * from t;
+----+------+
| pk | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.002 sec)
```

Client 1|Client2
-----------|-------
begin optimistic;|
update t set v=2 where pk=1;|
update t set v=3 where pk=2;|
| |update t set v=0 where pk=1;
commit;|

In the log file, it prints out `previous statement: update t set v=3 where pk=2"`, but that's not the statement which causes the conflict. It is more useful to print out `update t set v=2 where pk=1;`

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve? 

Fixes #14593 

### What is changed and how it works?

```
MySQL [test]> select * from t;
+----+------+
| pk | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.002 sec)
```

Client 1|Client2
-----------|-------
begin optimistic;|
update t set v=2 where pk=1;|
update t set v=3 where pk=2;|
| |update t set v=0 where pk=1;
commit;|

The result will be:

```
MySQL [test]> commit;
ERROR 9007 (HY000): Write conflict, txnStartTS=414420145598627841, conflictStartTS=414420147931709441, conflictCommitTS=414420147931709442, key={tableID=5262, handle=1} primary={tableID=5262, handle=1}, conflict cause: update t set v=2 where pk=1 [try again later]
```

In order to know which statement causes the conflict, we need to store the statement index together with the kv pair. So `MemBuffer` is modified to support storing **extra bytes* associated with keys. Here, the extra bytes are used to store the statement index. We might also enlarge the space and store something else.

Because ticlient does not have access to the statement history, we need to return the conflicting statement index to session. However, errors returned from ticlient don't carry data (only messages), so the conflicting statement index is just stored in `tikvTxn` and is fetched by the caller.

To add the statement text to error message, I have no other idea but use a dirty way. `parser/terror` is changed to expose the inner arguments. So I can change the messages in `ErrWriteConflict`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - This PR depends on history no matter whether retry is enabled, so the changes in #11192 is basically reverted. I'm not sure if the memory usage really counts.

### CODE REVIEW

File: executor/seqtest/seq_executor_test.go [hunk 1]
```diff
@@ -760,43 +760,6 @@ func (s *seqTestSuite) TestAdminShowNextID(c *C) {
 	r.Check(testkit.Rows("test1 tt id 41"))
 }
 
-func (s *seqTestSuite) TestNoHistoryWhenDisableRetry(c *C) {
```
Comment (jackysp) [inline executor/seqtest/seq_executor_test.go]: removed?

Comment (sticnarf) [reply executor/seqtest/seq_executor_test.go]: This feature requires history to be always saved. As mentioned in the PR description, #11192 is basically reverted.

**Comment by sticnarf** [comment]
Close this PR as the benefit does not match its cost.

**Pull request closed without merging**